hazenlib/ghosting.py

```diff
@@ -131,75 +132,95 @@ def get_eligible_area(signal_bounding_box, dcm, slice_radius=5):
+                          np.newaxis], np.array(
+                range(left_column, right_column))
 
     return eligible_columns, eligible_rows
 
 
-def get_ghost_slice(signal_bounding_box, dcm, slice_radius=5):
-    max_mean = 0
-    max_index = (0, 0)
-    windows = {}
-    arr = dcm.pixel_array
-
-    eligible_columns, eligible_rows = get_eligible_area(signal_bounding_box, dcm, slice_radius)
-    for idx, centre_voxel in np.ndenumerate(arr):
-        if idx[0] not in eligible_rows or idx[1] not in eligible_columns:
-            continue
-        else:
-            windows[idx] = arr[idx[0]-slice_radius:idx[0]+slice_radius, idx[1]-slice_radius:idx[1]+slice_radius]
 
-    for idx, window in windows.items():
-        if np.mean(window) > max_mean:
-            max_mean = np.mean(window)
-            max_index = idx
 
+
+def get_ghost_slice(signal_bounding_box, dcm, slice_radius=5):
+    eligible_area = get_eligible_area(signal_bounding_box, dcm, slice_radius=slice_radius)
     ghost_slice = np.array(
-        range(max_index[1] - slice_radius, max_index[1] + slice_radius), dtype=np.intp)[:, np.newaxis], np.array(
-        range(max_index[0] - slice_radius, max_index[0] + slice_radius)
+        range(min(eligible_area[1]),max(eligible_area[1])), dtype=np.intp)[:, np.newaxis], np.array(
+        range(min(eligible_area[0]), max(eligible_area[0]))
     )
-    #ghost_slice returns columns, rows of indexes
     return ghost_slice
 
 
-def get_ghosting(dcm, plotting=False) -> dict:
+
+
+
+
+
+
+def get_ghosting(dcm, report_path=False) -> dict:
 
     bbox = get_signal_bounding_box(dcm.pixel_array)
 
     x, y = hazenlib.get_pixel_size(dcm)  # assume square pixels i.e. x=y
     # ROIs need to be 10mmx10mm
     slice_radius = int(10//(2*x))
 
+
     signal_centre = [(bbox[0]+bbox[1])//2, (bbox[2]+bbox[3])//2]
     background_rois = get_background_rois(dcm, signal_centre)
     ghost_col, ghost_row = get_ghost_slice(bbox, dcm, slice_radius=slice_radius)
-    ghost = dcm.pixel_array[(ghost_row, ghost_col)]
+    ghost = dcm.pixel_array[(ghost_col, ghost_row)]
     signal_col, signal_row = get_signal_slice(bbox, slice_radius=slice_radius)
     phantom = dcm.pixel_array[(signal_row, signal_col)]
 
+
+
     noise = np.concatenate([dcm.pixel_array[(row, col)] for col, row in get_background_slices(background_rois, slice_radius=slice_radius)])
+
+
+
+
     eligible_area = get_eligible_area(bbox, dcm, slice_radius=slice_radius)
+
     ghosting = calculate_ghost_intensity(ghost, phantom, noise)
 
-    if plotting:
+    if report_path:
         import matplotlib.pyplot as plt
         fig, ax = plt.subplots()
         x1, x2, y1, y2 = bbox
 
-        img = hazenlib.rescale_to_byte(dcm.pixel_array)
+        img = dcm.pixel_array
+        img = img.astype('float64')
+        #print('this is img',img)
+        img *= 255.0 / img.max()
+        #img = hazenlib.rescale_to_byte(dcm.pixel_array)
```

Comment:
The `rescale_to_byte` function is also used in `spatial_resolution.py` and possibly others. I'm a little concerned by changing it here and not doing wholesale across the entire repo. What do you think @laurencejackson ?

Lu – can you upload some images of `img` generated using `rescale_to_byte` and your new version, to help illustrate how the data differs? I just want to get a better sense of what's changed.

Reply:
Hi Tom, the original rescale_to_byte function performs an histogram equalisation, which slightly manipulates the data. I think it is unnecessary to manipulate the data when we can just rescale in a more conventional way which does not manipulate the data. However, for this ghosting script, the rescale_to_byte is only used to plot, so actually I would leave the rescale_to_byte original function. I only changed this to check if it was making any difference while trying to fix the ghosting problem.